File: simulation/src/teamcode/control/path/Path.java
package teamcode.control.path;

import java.util.ArrayList;
import java.util.LinkedList;

import sim.company.ComputerDebugging;
import sim.company.FloatPoint;
import sim.company.Robot;
import teamcode.control.controllers.PurePursuitController;
import teamcode.control.path.PathPoints.*;
import teamcode.control.path.builders.PathBuilder;
import teamcode.util.MathUtil;
import teamcode.util.Point;
import teamcode.util.Pose;

public class Path extends LinkedList<PathPoints.BasePathPoint> {
    // target is always getFirst(), curr is copied
    public BasePathPoint curr;
    public boolean isPurePursuit;
    public boolean copi = true;
    public ArrayList<BasePathPoint> initialPoints;

    public String name;

    public Path(Path path) {
        for (BasePathPoint pathPoint : path) add(new BasePathPoint(pathPoint));

        initialPoints = new ArrayList<>();
        for(BasePathPoint pathPoint : this) initialPoints.add(new BasePathPoint(pathPoint));

        isPurePursuit = true;
        curr = new BasePathPoint(getFirst());
        removeFirst();

        name = path.name;
    }

    public Path(String name) {
        this.name = name;
    }

    public Path(BasePathPoint target, String name) {
        this(new PathBuilder(name).addPoint(target).build());
        isPurePursuit = false;
    }

    public void follow(Robot robot) {
        BasePathPoint target = getFirst();
        if(!isPurePursuit) {
//            PurePursuitController.goToPosition(robot, getFirst());
        } else {
            boolean skip;
//
//            if(size() > 1 && get(1).isStop != null && copi) {
//                Point cop = new Point(target);
//
//                getFirst().x = MathUtil.extendLine(curr, cop, getFirst().followDistance).x;
//                getFirst().y = MathUtil.extendLine(curr, cop, getFirst().followDistance).y;
//                System.out.println("EXTENDED: " +  getFirst());
//                System.out.println("copi: " + copi);
//                copi = false;
//            }

            if (target.isOnlyTurn != null) {
                skip = MathUtil.angleThresh(robot.currPose.heading, target.lockedHeading);
            } else if(target.isStop != null){
                skip = robot.currPose.distance(target) < 2; // test?
            } else {
                skip = robot.currPose.distance(target) < target.followDistance;
                if(size()>1 && get(1).isStop != null) {
                    skip = robot.currPose.distance(target) < 10;
                    System.out.println("ffffffffffffffffff");
                }
            }

            if(target.functions.size() != 0) {
                target.functions.removeIf(f -> f.cond() && f.func());
                skip = skip && target.functions.size() == 0;
            }

            if (skip) {
                curr = new BasePathPoint(target); // swap old target to curr start
                removeFirst();
            }

            if(finished()) return;
            if (target.isStop != null && robot.currPose.distance(target) < target.followDistance) {
                PurePursuitController.goToPosition(robot, target, target, curr);
                System.out.println("GOING TO FINAL STOP");
            }
            else {
                PurePursuitController.followPath(robot, curr, target, initialPoints);
                System.out.println("following curve");
            }
        }
    }


    @Override
    public String toString() {
        StringBuilder s = new StringBuilder("Path Name: " + name);
        String newLine = System.getProperty("line.separator");
        for(BasePathPoint p : initialPoints) {
            s.append(newLine).append("\t");

            if(curr.equals(p))
                s.append("curr:");
            else
                s.append("\t");
            s.append("\t");
            s.append(p);
        }
        s.append(newLine);
        return s.toString();
    }

    public boolean finished() {
        return size()==0;
    }

}

File: simulation/src/teamcode/control/Path.kt
package teamcode.control

import main.Azusa
import teamcode.control.path.funcs.Functions
import teamcode.control.waypoints.PointTurnWaypoint
import teamcode.control.waypoints.StopWaypoint
import teamcode.control.waypoints.Waypoint
import teamcode.util.Angle
import teamcode.util.AngleUnit
import teamcode.util.Point
import teamcode.util.Pose
import kotlin.collections.ArrayList

class Path(
        var waypoints: ArrayList<Waypoint>,
) {
    private var currWaypoint: Int
    private var interrupting: Boolean

    fun follow() {
        val currPose = Pose(Point(Azusa.worldXPosition, Azusa.worldYPosition), Angle(Azusa.worldAngleRad, AngleUnit.RAD))

        if (interrupting) {
            val advance = (waypoints[currWaypoint].func as Functions.InterruptFunction).run(this)
            if (advance)
                interrupting = false
            else return
        }

        var skip: Boolean
        do {
            val target = waypoints[currWaypoint + 1]

            skip = when (target) {
                is StopWaypoint -> currPose.distance(target) < 0.8
                is PointTurnWaypoint -> ((currPose.h - target.h).rad < target.dh.rad)
                else -> currPose.distance(target) < target.followDistance
            }

            var currAction = waypoints[currWaypoint].func
            if (currAction is Functions.RepeatFunction) {
                currAction.run(this)
            } else if (currAction is Functions.LoopUntilFunction) {
                skip = currAction.run(this)
            }

            if (skip) {
                currWaypoint++

                currAction = waypoints[currWaypoint].func
                if (currAction is Functions.SimpleFunction) {
                    currAction.run(this)
                }
                if (currAction is Functions.InterruptFunction) {
                    interrupting = true
                    this.follow()
                    return
                }
            }
        } while (skip && currWaypoint < waypoints.size - 1)
        if (finished()) return

        val target = waypoints[currWaypoint + 1]

        if (target is StopWaypoint && currPose.distance(target) < target.followDistance) {
            PurePursuitController.goToPosition(currPose, target)
        } else if (target is PointTurnWaypoint) {
            PurePursuitController.goToPosition(currPose, target)
        } else {
            PurePursuitController.followPath(waypoints[currWaypoint], target)
        }
    }

    fun finished() = currWaypoint >= waypoints.size - 1

    init {
        val copyList = ArrayList<Waypoint>()
        waypoints.forEach {copyList.add(it) }
        waypoints = copyList

        interrupting = false
        currWaypoint = 0
    }
}
